<a href="https://colab.research.google.com/github/Hritikrai55/CSAT-ANN-Project/blob/main/CSAT_ANN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Overview of Project**
This project focuses on predicting Customer Satisfaction (CSAT) scores using Deep Learning Artificial Neural Networks (ANN). In the context of e-commerce, understanding customer satisfaction through their interactions and feedback is crucial for enhancing service quality, customer retention, and overall business growth. By leveraging advanced neural network models, we aim to accurately forecast CSAT scores based on a myriad of interaction-related features, providing actionable insights for service improvement.

## About Dataset
The dataset captures customer satisfaction scores for a one-month period at an e-commerce platform called Shopzilla (a pseudonym). It includes various features such as category and sub-category of interaction, customer remarks, survey response date, category, item price, agent details (name, supervisor, manager), and CSAT score etc.

## **Project Background**
Customer satisfaction in the e-commerce sector is a pivotal metric that influences loyalty, repeat business, and word-of-mouth marketing. Traditionally, companies have relied on direct surveys to gauge customer satisfaction, which can be time-consuming and may not always capture the full spectrum of customer experiences. With the advent of deep learning, it's now possible to predict customer satisfaction scores in real-time, offering a granular view of service performance and identifying areas for immediate improvement.

## Data Description:

Unique id:	Unique identifier for each record

Channel name:	Name of the customer service channel

Category:	Category of the interaction

Sub-category:	Sub-category of the interaction

Customer Remarks:	Feedback provided by the customer

Order id:	Identifier for the order associated with the interaction

Order date time:	Date and time of the order

Issue reported at:	Timestamp when the issue was reported

Issue responded:	Timestamp when the issue was responded to

Survey response date:	Date of the customer survey response

Customer city:	City of the customer

Product category:	Category of the product

Item price:	Price of the item

Connected handling time:	Time taken to handle the interaction

Agent name:	Name of the customer service agent

Supervisor:	Name of the supervisor

Manager:	Name of the manager

Tenure Bucket:	Bucket categorizing agent tenure

Agent Shift:	Shift timing of the agent

CSAT Score:	Customer Satisfaction (CSAT) score

## **Project Goal**
The primary goal of this project is to develop a deep learning model that can accurately predict the CSAT scores based on customer interactions and feedback. By doing so, we aim to provide e-commerce businesses with a powerful tool to monitor and enhance customer satisfaction in real-time, thereby improving service quality and fostering customer loyalty.

In [ ]:
#Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")
import zipfile

from sklearn.preprocessing import FunctionTransformer
import scipy.stats as stats

In [ ]:
#Extract, read data and Create data frame
df=pd.read_csv('eCommerce_Customer_support_data.csv')

In [ ]:
df.head()

,Unique id,channel_name,category,Sub-category,Customer Remarks,Order_id,order_date_time,Issue_reported at,issue_responded,Survey_response_Date,Customer_City,Product_category,Item_price,connected_handling_time,Agent_name,Supervisor,Manager,Tenure Bucket,Agent Shift,CSAT Score
0,7e9ae164-6a8b-4521-a2d4-58f7c9fff13f,Outcall,Product Queries,Life Insurance,NaN,c27c9bb4-fa36-4140-9f1f-21009254ffdb,NaN,01/08/2023 11:13,01/08/2023 11:47,01-Aug-23,NaN,NaN,NaN,NaN,Richard Buchanan,Mason Gupta,Jennifer Nguyen,On Job Training,Morning,5
1,b07ec1b0-f376-43b6-86df-ec03da3b2e16,Outcall,Product Queries,Product Specific Information,NaN,d406b0c7-ce17-4654-b9de-f08d421254bd,NaN,01/08/2023 12:52,01/08/2023 12:54,01-Aug-23,NaN,NaN,NaN,NaN,Vicki Collins,Dylan Kim,Michael Lee,>90,Morning,5
2,200814dd-27c7-4149-ba2b-bd3af3092880,Inbound,Order Related,Installation/demo,NaN,c273368d-b961-44cb-beaf-62d6fd6c00d5,NaN,01/08/2023 20:16,01/08/2023 20:38,01-Aug-23,NaN,NaN,NaN,NaN,Duane Norman,Jackson Park,William Kim,On Job Training,Evening,5
3,eb0d3e53-c1ca-42d3-8486-e42c8d622135,Inbound,Returns,Reverse Pickup Enquiry,NaN,5aed0059-55a4-4ec6-bb54-97942092020a,NaN,01/08/2023 20:56,01/08/2023 21:16,01-Aug-23,NaN,NaN,NaN,NaN,Patrick Flores,Olivia Wang,John Smith,>90,Evening,5
4,ba903143-1e54-406c-b969-46c52f92e5df,Inbound,Cancellation,Not Needed,NaN,e8bed5a9-6933-4aff-9dc6-ccefd7dcde59,NaN,01/08/2023 10:30,01/08/2023 10:32,01-Aug-23,NaN,NaN,NaN,NaN,Christopher Sanchez,Austin Johnson,Michael Lee,0-30,Morning,5


In [ ]:
print("This data set has",df.shape[0],"rows and",df.shape[1],"columns.")

This data set has 85907 rows and 20 columns.


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85907 entries, 0 to 85906
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Unique id                85907 non-null  object 
 1   channel_name             85907 non-null  object 
 2   category                 85907 non-null  object 
 3   Sub-category             85907 non-null  object 
 4   Customer Remarks         28742 non-null  object 
 5   Order_id                 67675 non-null  object 
 6   order_date_time          17214 non-null  object 
 7   Issue_reported at        85907 non-null  object 
 8   issue_responded          85907 non-null  object 
 9   Survey_response_Date     85907 non-null  object 
 10  Customer_City            17079 non-null  object 
 11  Product_category         17196 non-null  object 
 12  Item_price               17206 non-null  float64
 13  connected_handling_time  242 non-null    float64
 14  Agent_name            

Based on data set info. This df has dumped with 85907 rows and 20 Columns.

Except 3 variable such as (Item_price, Connected_handling_time and CSAT Score) all other variables seems to be Object data type.

The data shows like non-null but as per count in some variable, many data might be fill with NAN, NA or some other charecters.

In [ ]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Item_price,17206.0,5660.774846,12825.728411,0.0,392.0,979.0,2699.75,164999.0
connected_handling_time,242.0,462.400826,246.295037,0.0,293.0,427.0,592.25,1986.0
CSAT Score,85907.0,4.242157,1.378903,1.0,4.0,5.0,5.00,5.0


Describe is actually an 5 point summary which will calculate for numerical values. So the reason only 3 variable got described.

* Item_price and connected_handling_time has its bell shape curve with a huge left skewness and these both seems to have too many outlier.

* CSAT variable has numerical Dtype but it looks like a categorical variable.


In [ ]:
for i in df.columns:
    print(df[i].value_counts())

Unique id
07c7a878-0d5a-42e0-97ef-de59abec0238    1
7e9ae164-6a8b-4521-a2d4-58f7c9fff13f    1
b07ec1b0-f376-43b6-86df-ec03da3b2e16    1
c0d72065-e4c8-449b-9d55-cadb690b9c54    1
49e1675a-e890-4d27-93a0-51e67a44d86a    1
                                       ..
372b51a5-fa19-4a31-a4b8-a21de117d75e    1
11a3ffd8-1d6b-4806-b198-c60b5934c9bc    1
1cfde5b9-6112-44fc-8f3b-892196137a62    1
ba903143-1e54-406c-b969-46c52f92e5df    1
eb0d3e53-c1ca-42d3-8486-e42c8d622135    1
Name: count, Length: 85907, dtype: int64
channel_name
Inbound    68142
Outcall    14742
Email       3023
Name: count, dtype: int64
category
Returns               44097
Order Related         23215
Refund Related         4550
Product Queries        3692
Shopzilla Related      2792
Payments related       2327
Feedback               2294
Cancellation           2212
Offers & Cashback       480
Others                   99
App/website              84
Onboarding related       65
Name: count, dtype: int64
Sub-category
Reverse Picku

In [ ]:
#finding duplicate datas
df.duplicated().sum()

np.int64(0)

In [ ]:
# finding null values and unique value counts
null_values= round(df.isnull().sum() * 100 / len(df),2)
unique_values=df.nunique()
analysis_df = pd.DataFrame({'column_name': df.columns,"unique_values": unique_values,'null_values_%': null_values,"data_type":df.dtypes})
analysis_df.reset_index(drop=True,inplace=True)
analysis_df


,column_name,unique_values,null_values_%,data_type
0,Unique id,85907,0.00,object
1,channel_name,3,0.00,object
2,category,12,0.00,object
3,Sub-category,57,0.00,object
4,Customer Remarks,18231,66.54,object
5,Order_id,67675,21.22,object
6,order_date_time,13766,79.96,object
7,Issue_reported at,30923,0.00,object
8,issue_responded,30262,0.00,object
9,Survey_response_Date,31,0.00,object
